### PR TITLE
Adds the Kollect free implementation with some TODO parts

### PR DESCRIPTION
**Pull request by fedefernandez**
I've been stuck in the `TODO` parts for several days, so I've decided to create the PR and tackle them in other PRs.

Main changes:
* `KollectRequest` and `KollectQuery` are now interfaces and `KollectOp` a `sealed class`
* `KollectOp` has been manually implemented as a higher kind instead of using `@higherkind`

I didn't fully assimilate the Kotlin language and the Arrow constructors yet, so any feedback is welcomed.

Thanks

### CODE REVIEW

**Comment by fedefernandez** [comment]
@raulraja I've finished all the parts except the "runners". As we talked the other day, we'll implement the runners as extension methods for the applicative instance in another PR. So, this is ready for review. Thanks!

File: kollect-core/src/main/kotlin/kollect/Kollect.kt [hunk 1]
```diff
@@ -30,20 +34,129 @@ sealed class KollectQuery<I : Any, A> : KollectRequest() {
...
+         * Apply a kollect-returning function to every element in a list and return a Kollect of the list of
+         * results. It implies concurrent execution of kollects.
+         */
+        fun <A, B> traverse(ids: List<A>, f: (A) -> Kollect<B>): Kollect<List<B>> =
+                traverseGrouped(ids, 50, f)
+
+        fun <A> ListK<A>.grouped(n: Int): ListK<ListK<A>> =
+            withIndex()
+                    .groupBy { it.index / n }
+                    .map { it.value.k().map { it.value } }.k()
```
Comment (peterneyens) [inline kollect-core/src/main/kotlin/kollect/Kollect.kt]: Apparently my previous comment got lost somewhere.

I am not sure if I understand the last line 100%, but I think this `grouped` implementation will result in a `traverse` that returns the results in a different order than the input.
It seems to me that `(1,2,3,4,5)` with `grouped(2)` would become `((1,3,5), (2,4))`?

Comment (fedefernandez) [reply kollect-core/src/main/kotlin/kollect/Kollect.kt]: I can't assure that it's the best implementation but it works for me. You can try it by yourself at [try.kotlinlang.org](https://try.kotlinlang.org/#/UserProjects/str02igkvqj1tevjid3pb0d421/42bcgmd91f4b673helqqvbv6et) 

```kotlin
fun main(args: Array<String>) {
    fun <A> List<A>.grouped(n: Int): List<List<A>> =
            withIndex()
                    .groupBy { it.index / n }
                    .map { it.value.map { it.value } }

	val a = listOf(1,2,3,4,5)

	println(a.grouped(2))
}
```

Comment (peterneyens) [reply kollect-core/src/main/kotlin/kollect/Kollect.kt]: Yes, you are right, `it.index % n` would do what I mentioned before.
We can always come back and improve this implementation.

**Pull request merged**